pf/internal/defaults/defaults.go

```diff
@@ -0,0 +1,293 @@
+	switch sch.Type() {
+	case shim.TypeBool:
+		v := false
+		if str != "" {
+			if v, err = strconv.ParseBool(str); err != nil {
+				return resource.NewNullProperty(), err
+			}
+		}
+		return resource.NewBoolProperty(v), nil
+	case shim.TypeInt:
+		v := int(0)
+		if str != "" {
+			iv, iverr := strconv.ParseInt(str, 0, 0)
+			if iverr != nil {
+				return resource.NewNullProperty(), iverr
+			}
+			v = int(iv)
+		}
+		return resource.NewNumberProperty(float64(v)), nil
+	case shim.TypeFloat:
+		v := float64(0.0)
+		if str != "" {
+			if v, err = strconv.ParseFloat(str, 64); err != nil {
+				return resource.NewNullProperty(), err
+			}
+		}
+		return resource.NewNumberProperty(v), nil
+	case shim.TypeString:
+		return resource.NewStringProperty(str), nil
+	default:
+		return resource.NewNullProperty(), fmt.Errorf("unknown type for default value: %v", sch.Type())
+	}
+}
+
+func recoverDefaultValue(defaultValue any) resource.PropertyValue {
+	if pv, alreadyPV := defaultValue.(resource.PropertyValue); alreadyPV {
+		return pv
+	}
+	return resource.NewPropertyValue(defaultValue)
+}
+
+type defaultsTransform struct {
+	topSchemaMap     shim.SchemaMap
+	topFieldInfos    map[string]*tfbridge.SchemaInfo // optional
+	resourceInstance *tfbridge.PulumiResource        // optional
+	providerConfig   resource.PropertyMap            // optional
+}
+
+// Returns a non-nil resourceInstance only if the defaults are being applied to a resource at the top level.
+func (du *defaultsTransform) resourceByPath(path resource.PropertyPath) *tfbridge.PulumiResource {
+	var res *tfbridge.PulumiResource
+	if len(path) == 0 {
+		res = du.resourceInstance
+	}
+	return res
+}
+
+// Returns matching object schema for a context determined by the PropertyPath, if any.
+func (du *defaultsTransform) lookupSchemaByContext(
+	path resource.PropertyPath,
+) (shim.SchemaMap, map[string]*tfbridge.SchemaInfo, bool) {
+	if len(path) == 0 {
+		return du.topSchemaMap, du.topFieldInfos, true
+	}
+
+	schemaPath := tfbridge.PropertyPathToSchemaPath(path, du.topSchemaMap, du.topFieldInfos)
+	if schemaPath == nil {
+		return nil, nil, false
+	}
+
+	schema, info, err := tfbridge.LookupSchemas(schemaPath, du.topSchemaMap, du.topFieldInfos)
+	if err != nil {
+		return nil, nil, false
+	}
+
+	encodedObjectSchema, ok := schema.Elem().(shim.Resource)
+	if !ok {
+		return nil, nil, false
+	}
+
+	objectSchema := encodedObjectSchema.Schema()
+
+	var fields map[string]*tfbridge.SchemaInfo
+	if info == nil || info.Fields == nil {
+		fields = map[string]*tfbridge.SchemaInfo{}
+	} else {
+		fields = info.Fields
+	}
+
+	return objectSchema, fields, true
+}
+
+// Extends PropertyMap with Pulumi-specified default values.
+func (du *defaultsTransform) extendPropertyMapWithDefaults(
+	ctx context.Context,
+	path resource.PropertyPath,
+	props resource.PropertyMap,
+) (resource.PropertyMap, error) {
+	schemaMap, infos, ok := du.lookupSchemaByContext(path)
+	if !ok {
+		return props, nil
+	}
+	res := props.Copy()
```

Comment:
This is not a deep copy. It looks like this is fine, but I wanted to make sure it wasn't relied upon.

https://github.com/pulumi/pulumi/blob/ae247a81951b1eeb8b43500fdc13cc95d16be433/sdk/go/common/resource/properties.go#L195-L203

Reply:
Yeah, it's fine, we want exactly a shallow copy here, and we're filling out missing fields in an object. The deep recursion is taken in propertyvalue.TransformPropertyValue so this can stay at one level.